### PR TITLE
Add Integration tests

### PR DESCRIPTION
**Pull request by beccaelenzil**
Updates for C17 to align with our focus on writing test:
- Integration test for Waves 01, 02, 03
- Integration test for Wave 04, 05, 06
- Remove some competed tests from Wave 01, 02, and 06 (completed_unit_tests directory won't be included on master branch)
- Update workflow in README to emphasize TDD
- Add Integration Test section in README
- Add code coverage section in README
- Update details on running tests section
- Add writing unit tests to optional feature

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -87,29 +87,35 @@ Summary of one-time project setup:
...
-2. Find the test file that contains the test you want to run.
+2. Check the `tests/unit_tests` folder, and find the test file you want to run
 
-   - Check the `tests` folder, and find the test file you want to run
    - In that test file, read through each test case
+      - If it is incomplete, complete the test. **There are incomplete tests for Waves 01, 02, and 06**. 
+        - *Is this a nominal or edge case?*
+        - *What type of input do we need to test this case?*
+        - *What is the expected output for the given input?*
+   - Remove the lines that contain `@pytest.mark.skip()` for the test(s) you want to run.
```
Comment (kp483) [inline README.md:97]: It doesn't seem like `@pytest.mark.skip()` is used for any tests in this project, so could be omitted unless this is just general information included for all projects

**Review by kp483 — APPROVED** [review]
Looks good! Just a couple minor suggestions/requests for clarification in the readme. Integration tests look good!

File: README.md (hunk 1 above)
Comment (beccaelenzil) [reply README.md:97]: Oh! Good catch. I was thinking I'd add these to each test. Though open to leaving them off. I know they are in viewing party for each test.

File: README.md [hunk 2]
```diff
@@ -87,29 +87,35 @@ Summary of one-time project setup:
...
 
-3. Run the tests for your specific wave
+3. Run the test(s)! (RED)
```
[the remark this answers is not in the document]
Comment (beccaelenzil) [reply README.md:101]: I meant to add this line to the top of the project workflow "We will use a Test Driven Development programming workflow to work on this project. Notice the Red-Green-Refactor steps in the workflow steps outlined below.", and then add (GREEN) to step 6

File: README.md [hunk 3]
```diff
@@ -87,29 +87,35 @@ Summary of one-time project setup:
...
 
-4. Focus on the top test failure. Read through the test failure, and understand why the failure is happening. Confirm your findings with a classmate.
+4. Read through the test failure, and understand why the failure is happening. Confirm your findings with a classmate. 
+    - If it is a test you wrote, consider whether you are actually testing what you intend to test. Does the test need modification?
 
-5. Make a plan to fix the test failure.
+5. Make a plan to implement code to pass the test.
 
-6. Write code to fix the test failure.
+6. Write code to pass the test.
```
Comment (beccaelenzil) [inline README.md]:
```suggestion
6. Write code to pass the test. (GREEN)
```

File: README.md (hunk 1 above)
Comment (kelsey-steven-ada) [reply README.md:99]: If students won't have seen this before, would it be worth adding something that reminds them to search the project/test files for `@pytest.mark.skip()` to ensure no tests were missed before submitting the assignment?

File: README.md [hunk 4]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+7. Re-run the test(s).
 
-7. Re-run the tests.
+8. Repeat steps 3-7 until that test passes!
 
-8. Repeat steps 5-7 until that test passes!
+9. Consider looking for opportunities to improve your code (REFACTOR)
 
-9. Repeats steps 4-8 until you have finished all tests in the file.
+10. Repeats steps 2-8 until you have finished all tests in the file.
```
Comment (kelsey-steven-ada) [inline README.md:122]: Couple small things:
- Should it be steps 2-9 to include the refactor step?
- I think "Repeats" should be "Repeat"

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply README.md:101]: I'm not familiar enough with our curriculum to know the answer to this, but will students have seen this notation before? I know a lot of folks will probably immediately get red is for test should be not passing here, but my instinct is this could raise confusion if they haven't definitely seen it before, even with the paragraph above.

File: README.md [hunk 5]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
 ```bash
-# Must be in activated virtual environment
-$ pytest tests/test_wave_01.py
+# Must be in activated virtual environment in the project-root directory
+$ pytest
 ```
 
-4. Focus on the top test failure. Read through the test failure, and understand why the failure is happening. Confirm your findings with a classmate.
+4. Read through the test failure, and understand why the failure is happening. Confirm your findings with a classmate. 
+    - If it is a test you wrote, consider whether you are actually testing what you intend to test. Does the test need modification?
```
Comment (kelsey-steven-ada) [inline README.md:110]: Feel free to ignore, I just had the thought that this tweak might focus on what the code is doing.

```suggestion
    - If it is a test you wrote, consider whether the code is actually testing what you intend to test. Does the test need modification?
```

File: README.md [hunk 6]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+9. Consider looking for opportunities to improve your code (REFACTOR)
 
-9. Repeats steps 4-8 until you have finished all tests in the file.
+10. Repeats steps 2-8 until you have finished all tests in the file.
 
-10. Begin using the test file of the next wave!
+11. Begin using the test file of the next wave!
 
-11. When you are finished working for the day, deactivate your environment with deactivate or closing the Terminal tab/window
+12. When you are finished working for the day, deactivate your environment with deactivate or closing the Terminal tab/window
```
Comment (kelsey-steven-ada) [inline README.md:126]: Small grammar suggestion

```suggestion
12. When you are finished working for the day, deactivate your environment with deactivate or by closing the Terminal tab/window
```

File: README.md [hunk 7]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
 
 ```bash
 $ deactivate
 ```
+Finally, at submission time, **no matter where you are**, submit the project via Learn.
```
Comment (kelsey-steven-ada) [inline README.md:131]: I like adding the emphasis here to turn in the project wherever it may be.

File: README.md [hunk 8]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+
+This will let us give feedback on what you've finished so that you can be better prepared for the next project.
+
+## Integration Tests
+
+We provided integration tests for this project. Integration tests verify that individual units of a piece of software are working together correctly.
```
Comment (kelsey-steven-ada) [inline README.md:137]: Is there a way to reword this that helps students understand how the integration tests are distinct from unit tests/the other tests in the project? The paragraph below definitely implies that they are more involved than unit tests, but my thinking is more clarity could be helpful.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (beccaelenzil) [reply README.md:101]: Yea, I think you're right that it adds more confusion than value. I'll remove it here and from the viewing party README.

File: README.md [hunk 9]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+
+The integration tests for this project are in the directory `tests/integration_tests`, and the tests have the decorator `@pytest.mark.integration_test` provided by the `pytest-integration` package. Marking these tests as integration tests makes them run after the unit tests. The isolated nature and specficity of *unit tests* make them a much better tool for debugging. Hence, we want to run the unit tests before the integration tests.
+
+The integration tests provided in this project do not cover all the test cases verified by the unit tests. They verify that the functions work together when invoked correctly. We could classify them as the subset of Integration Tests called *Smoke Tests*, tests that check basic functionality of the application.
```
Comment (kelsey-steven-ada) [inline README.md:141]: This paragraph really gets at what I meant with my prior comment on line 37. Would it make sense to move this up as part of that first paragraph that introduces what the integration tests are?

File: README.md [hunk 10]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+
+For more on different types of software testing, [here is a helpful resource](https://www.atlassian.com/continuous-delivery/software-testing/types-of-software-testing). 
+
+## Code Coverage
+
+It is good practice to check code coverage. Code coverage is a term that is used to describe how much application code is executed when a particular test suite is run. A test suite with a high percentage of coverage is likely to be testing more throughly and have fewer bugs. A code coverage tool can partner with our testing suite to give us a report illustrating the coverage of our tests.
```
Comment (kelsey-steven-ada) [inline README.md:147]: I was thinking it might be helpful to define code coverage first and provide a little more "why" in the good practice sentence. 

```suggestion
Code coverage is a term used to describe how much application code is executed when a particular test suite is run. It is a good practice to check our code coverage, to understand how much of our code is exercised by tests vs how much is still untested. A test suite with a high percentage of coverage is likely to be testing more throughly and have fewer bugs. A code coverage tool can partner with our testing suite to give us a report illustrating the coverage of our tests.
```

File: README.md [hunk 11]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+
+To run all unskipped tests that exist in this project with:
 
 ```bash
 # Must be in activated virtual environment
 $ pytest
 ```
 
-If you want to run all tests that exist in one file, use:
+To see any `print` statements print to the console, add `-s` to the end of any `pytest` command:
```
Comment (kelsey-steven-ada) [inline README.md:169]: Minor suggestion, not necessary, but this reads a little more naturally to me:

```suggestion
To see any `print` statement's output in the console, add `-s` to the end of any `pytest` command:
```

File: README.md [hunk 12]
```diff
@@ -81,69 +81,116 @@ Summary of one-time project setup:
...
+
+```bash
+# Must be in activated virtual environment
+$ pytest tests/unit_tests/test_file_name.py
 ```
 
 ... where `test_file_name.py` is replaced with the correct test file name.
 
-If you want to see any `print` statements print to the console, add `-s` to the end of any `pytest` command:
+To run a single test by name:
```
Comment (kelsey-steven-ada) [inline README.md:185]: I like the gentle reorganization here and adding in running a single test by name.

File: swap_meet/clothing.py [hunk 13]
```diff
@@ -1,2 +1,2 @@
 class Clothing:
-    pass
+    pass
```
Comment (kelsey-steven-ada) [inline swap_meet/clothing.py:2]: This might just be githib being over cautious, but do .py files require a new line at the end to be valid? On the diff I see this red icon with hover text on clothing.py, decor.py, and vendor.py but not on item.py which has several new lines in it.

Screen Shot 2022-03-04 at 2 35 39 PM

File: tests/integration_tests/test_wave_01_02_03.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    result2 = vendor.add(item2)
+
+    assert len(vendor.inventory) == 2
+    assert item1 in vendor.inventory
+    assert item2 in vendor.inventory
+    assert result1 == item1
+    assert result2 == item2
+
+    # remove an item
+    result = vendor.remove(item1)
```
Comment (kelsey-steven-ada) [inline tests/integration_tests/test_wave_01_02_03.py:24]: Since we have `result1` and `result2` above, does it make sense for this to be something like `result3` or `remove_result`?

File: tests/integration_tests/test_wave_04_05_06.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    camila.add(item_clothing1)
+    camila.add(item_clothing2)
+
+    valentina.add(item_electronics2)
+    valentina.add(item_decor1)
+    valentina.add(item_decor2)
+    
+
+    # swap first item
+    result = camila.swap_first_item(valentina)
```
Comment (kelsey-steven-ada) [inline tests/integration_tests/test_wave_04_05_06.py:29]: I know in test files in practice naming conventions are often a little more lax, but I was wondering if it's worth switching this to something like `swap_result` to more closely describe what it's holding?

File: tests/unit_tests/test_wave_06.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        their_priority="Decor"
+    )
+
+    # *********************************************************************
+    # ****** Complete Assert Portion of this test **********
+    # *********************************************************************
+    # Assertions should check:
+    # - That the results is truthy
+    # - That tai and jesse's inventories are the correct length
+    # - That all the correct items are in tai and jesse's inventories, and that the items that were swapped are not there
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_06.py:82]: It's a reach, but I had the thought that the tail end of the sentence could possibly give the impression those items shouldn't be in either vendor, feel free to ignore if it doesn't feel necessary. 

```suggestion
    # - That all the correct items are in tai and jesse's inventories, including the items which were swapped from one vendor to the other
```

**Comment by kelsey-steven-ada** [comment]
I reviewed the README on master before the PR and caught a small typo that I think is worth fixing while making changes. Under Wave 1 of Project Directions, there's the bullet: 
> When we create initialize an instance of Vendor, we can optionally pass in a list with the keyword argument inventory
that could use either `create` or `initialize` removed.

**Review by kelsey-steven-ada — APPROVED** [review]

File: README.md (hunk 9 above)
Comment (beccaelenzil) [reply README.md:141]: Good call. I took a stab at reorganizing.

File: swap_meet/clothing.py (hunk 13 above)
Comment (beccaelenzil) [reply swap_meet/clothing.py:2]: I don't think so. I think this is just because I had the solutions in there, accidentally committed them, and then deleted them. I removed the extra new lines.

File: README.md (hunk 8 above)
Comment (beccaelenzil) [reply README.md]: I took a stab at this reorganization. Would love feedback on this!

File: README.md (hunk 9 above)
Comment (kelsey-steven-ada) [reply README.md]: I think I'm stuck on the sentence "Integration tests verify that individual units of a piece of software are working together correctly." To me, it doesn't add as much clarity as I would like.  

I was thinking something that compared them directly to unit tests like:
> While unit tests are small, fast, and should cover most of our program's behavior down to individual functions, integration tests verify that the various pieces of a program are working together correctly.

File: tests/unit_tests/test_wave_06.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+        other=jesse,
+        my_priority="Clothing",
+        their_priority="Decor"
+    )
+
+    # *********************************************************************
+    # ****** Complete Assert Portion of this test **********
+    # *********************************************************************
+    # Assertions should check:
+    # - That the results is truthy
```
Comment (kelsey-steven-ada) [inline tests/unit_tests/test_wave_06.py:80]: Teeny grammar nitpick: should this be "That result is truthy" if result here is referring to the named variable `result`?

**Review by kelsey-steven-ada — APPROVED** [review]
Looks good, approving with a couple minor questions/suggestions.

File: README.md (hunk 9 above)
Comment (kp483) [reply README.md]: I like Kelsey's suggestion. I think specifically, using the word 'units' in the definition of integration tests might undermine the contrast we're trying to draw between unit and integration tests.

File: README.md (hunk 1 above)
Comment (beccaelenzil) [reply README.md]: Added @pytest.mark.skip to all tests and added a note to search for this in the readme (actually viewing-party readme, which this readme now points to)

File: tests/unit_tests/test_wave_06.py (hunk 17 above)
Comment (beccaelenzil) [reply tests/unit_tests/test_wave_06.py:83]: good call. fixed!

File: README.md (hunk 9 above)
Comment (beccaelenzil) [reply README.md]: great! I incorporated this.

**Comment by beccaelenzil** [comment]
I incorporated all the feedback, added @pytest.mark.skip to all the tests and verified they are all skip, and took out a huge chunk of the readme that is an exact copy of the viewing party readme. I think it would be helpful to have another set of eyes on the removal of the readme sections that now direct students to the viewing party readme.
